Template plugin class doc (#16152)
openvinotoolkit/openvino: docs/IE_PLUGIN_DG/Plugin.md

The provided plugin class also has several fields:

> * `m_backend` - a backend engine that is used to perform actual computations for network inference. For `Template` plugin `ov::runtime::Backend` is used which performs computations using OpenVINO™ reference implementations.

Reviewer: ```suggestion
* `m_backend` - a backend engine that is used to perform actual computations for model inference. For `Template` plugin `ov::runtime::Backend` is used which performs computations using OpenVINO™ reference implementations.
```

* `m_waitExecutor` - a task executor that waits for a response from a device about device tasks completion.
* `m_cfg` of type `Configuration`:

@snippet template/src/config.hpp configuration:header

As an example, a plugin configuration has three value parameters:

- `device_id` - particular device ID to work with. Applicable if a plugin supports more than one `Template` device. In this case, some plugin methods, like `set_property`, `query_model`, and `compile_model`, must support the ov::device::id property. 
- `perf_counts` - boolean value to identify whether to collect performance counters during [Inference Request](@ref openvino_docs_ie_plugin_dg_infer_request) execution.

> - `streams_executor_config` - configuration of `ov::threading::IStreamsExecutor` to handle settings of multi-threaded context.

Reviewer: perf_mode is missing here

### Plugin Constructor

A plugin constructor must contain code that checks the ability to work with a device of the `Template` 
type. For example, if some drivers are required, the code must check 
driver availability. If a driver is not available (for example, OpenCL runtime is not installed in 
case of a GPU device or there is an improper version of a driver is on a host machine), an exception 
must be thrown from a plugin constructor.

A plugin must define a device name enabled via the `set_device_name()` method of a base class:

@snippet template/src/plugin.cpp plugin:ctor

### Plugin Destructor

A plugin destructor must stop all plugins activities, and clean all allocated resources.

@snippet template/src/plugin.cpp plugin:dtor

### `compile_model()`

> The plugin should implement two `compile_model()` methods the first one compiles model without remote context, the second one with remote context if plugin supports.

Reviewer: ```suggestion
The plugin should implement two `compile_model()` methods: the first one compiles model without remote context, the second one with remote context if plugin supports.
```

> This is the most important function of the `Plugin` class and creates an instance of compiled `CompiledModel`,

Reviewer: ```suggestion
This is the most important function of the `Plugin` class is to create an instance of compiled `CompiledModel`,
```

which holds a backend-dependent compiled graph in an internal representation:

@snippet template/src/plugin.cpp plugin:compile_model

@snippet template/src/plugin.cpp plugin:compile_model_with_remote

Before a creation of an `CompiledModel` instance via a constructor, a plugin may check if a provided 
ov::Model object is supported by a device if it is needed.

> Actual graph compilation is done in the `CompiledModel` constructor. Refer to the [CompiledModel Implementation Guide](@ref openvino_docs_ie_plugin_dg_executable_network) for details.

Reviewer: ```suggestion
Actual model compilation is done in the `CompiledModel` constructor. Refer to the [CompiledModel Implementation Guide](@ref openvino_docs_ie_plugin_dg_executable_network) for details.
```

> **NOTE**: Actual configuration map used in `CompiledModel` is constructed as a base plugin 
> configuration set via `Plugin::set_property`, where some values are overwritten with `config` passed to `Plugin::compile_model`. 
> Therefore, the config of  `Plugin::compile_model` has a higher priority.

### `transform_model()`

> The function accepts a const shared pointer to `ov::Model` object and applies common and plugin-specific transformations on a copied graph to make the graph more friendly to hardware operations. For details how to write custom plugin-specific transformation, please, refer to [Writing OpenVINO™ transformations](@ref openvino_docs_transformations) guide. See detailed topics about network representation:

Reviewer: ```suggestion
The function accepts a const shared pointer to `ov::Model` object and applies common and device-specific transformations on a copied model to make it more friendly to hardware operations. For details how to write custom device-specific transformation, please, refer to [Writing OpenVINO™ transformations](@ref openvino_docs_transformations) guide. See detailed topics about network representation:
```

    * [Intermediate Representation and Operation Sets](@ref openvino_docs_MO_DG_IR_and_opsets)
    * [Quantized models](@ref openvino_docs_ie_plugin_dg_quantized_networks).

@snippet template/src/plugin.cpp plugin:transform_model

> > **NOTE**: After all these transformations, a `ov::Model` object contains operations which can be perfectly mapped to backend kernels. E.g. if backend has kernel computing `A + B` operations at once, the `transform_model` function should contain a pass which fuses operations `A` and `B` into a single custom operation `A + B` which fits backend kernels set.

Reviewer: ```suggestion
> **NOTE**: After all these transformations, an `ov::Model` object contains operations which can be perfectly mapped to backend kernels. E.g. if backend has kernel computing `A + B` operations at once, the `transform_model` function should contain a pass which fuses operations `A` and `B` into a single custom operation `A + B` which fits backend kernels set.
```

> ### `query_model()`

Reviewer: for some reason I don't see this header in the docs, as well as transform_model, compile_model and others

Use the method with the `HETERO` mode, which allows to distribute network execution between different 
devices based on the `ov::Node::get_rt_info()` map, which can contain the `"affinity"` key.
The `query_model` method analyzes operations of provided `model` and returns a list of supported

> operations via the ov::SupportedOpsMap structure. The `query_model` firstly applies `transform_model` passes to input `ov::Model` argument. After this, the transformed network in ideal case contains only operations are 1:1 mapped to kernels in computational backend. In this case, it's very easy to analyze which operations is supposed (`m_backend` has a kernel for such operation or extensions for the operation is provided) and not supported (kernel is missed in `m_backend`):

Reviewer: ```suggestion
operations via the ov::SupportedOpsMap structure. The `query_model` firstly applies `transform_model` passes to input `ov::Model` argument. After this, the transformed model in ideal case contains only operations are 1:1 mapped to kernels in computational backend. In this case, it's very easy to analyze which operations is supposed (`m_backend` has a kernel for such operation or extensions for the operation is provided) and not supported (kernel is missed in `m_backend`):
```

1. Store original names of all operations in input `ov::Model`
2. Apply `transform_model` passes. Note, the names of operations in a transformed network can be different and we need to restore the mapping in the steps below.
3. Construct `supported` map which contains names of original operations. Note, that since the inference is performed using OpenVINO™ reference backend, the decision whether the operation is supported or not depends on whether the latest OpenVINO opset contains such operation.
4. `ov.SupportedOpsMap` contains only operations which are fully supported by `m_backend`.

> @snippet template/src/plugin.cpp plugin:query_model

Reviewer: I see `InferenceEngine::GetSupportedNodes` is used in the snippet. It needs to be moved to new Plugin API.
Let's grep by IE / InferenceEngine in the template plugin code

### `set_property()`

Sets new values for plugin property keys:

> @snippet template/src/plugin.cpp plugin:set_property

Reviewer: it contains support of old metrics like SUPPORTED_METRICS and IMPORT_EXPORT_SUPPORT, can we drop it?

Reviewer: why?

In the snippet above, the `Configuration` class overrides previous configuration values with the new 
ones. All these values are used during backend specific graph compilation and execution of inference requests.

> **NOTE**: The function must throw an exception if it receives an unsupported configuration key.

### `set_property()`

Returns a current value for a specified property key:

@snippet template/src/plugin.cpp plugin:get_property

The function is implemented with the `Configuration::Get` method, which wraps an actual configuration 
key value to the ov::Any and returns it.

> **NOTE**: The function must throw an exception if it receives an unsupported configuration key.

### `import_model()`

> The importing network mechanism allows to import a previously exported backend specific graph and wrap it 

Reviewer: ```suggestion
The importing of compiled model mechanism allows to import a previously exported backend specific graph and wrap it 
```

using an [CompiledModel](@ref openvino_docs_ie_plugin_dg_executable_network) object. This functionality is useful if 
backend specific graph compilation takes significant time and/or cannot be done on a target host 
device due to other reasons.

During export of backend specific graph using `CompiledModel::export_model`, a plugin may export any 
type of information it needs to import a compiled graph properly and check its correctness. 
For example, the export information may include:

- Compilation options (state of `Plugin::m_cfg` structure)
- Information about a plugin and a device type to check this information later during the import and 
throw an exception if the `model` stream contains wrong data. For example, if devices have different 
capabilities and a graph compiled for a particular device cannot be used for another, such type of 
information must be stored and checked during the import. 
- Compiled backend specific graph itself

> - Information about precisions and shapes set by the user

Reviewer: I suppose it's not imported / exported anymore
